### PR TITLE
Feature Flags

### LINKED ISSUE #249: Feature: Selecting test cases based on available hardware/fixtures
To make the test environment really scalable, it would be great, if I could define for each test case, which dependencies it has. For example, when having a complete rack with switchable power supplies, IO boards and externel peripherals, I could run all tests. When running the tests on the desk, without having the full setup, I would define in a config, which HW I have and hereby filter the tests, that are executed (automatically).

### PR DESCRIPTION
**Pull request by Emantor**
- [x] Docs
- [x] Tests

Fixes #249

### CODE REVIEW

File: labgrid/resource/udev.py [hunk 1]
```diff
@@ -292,44 +292,24 @@ class USBSDMuxDevice(USBResource):
     def __attrs_post_init__(self):
         self.match['ID_VENDOR_ID'] = '0424'
         self.match['ID_MODEL_ID'] = '4041'
+        self.control_path = None
```
Comment (jluebbe) [inline labgrid/resource/udev.py]: This commit seems to be in the wrong PR.

File: labgrid/config.py [hunk 2]
```diff
@@ -196,3 +196,6 @@ def get_images(self):
             images[name] = self.resolve_path(image)
 
         return images
+
+    def get_feature_flags(self):
+        return self.data.get('feature flags', [])
```
Comment (jluebbe) [inline labgrid/config.py]: No spaces here, just use 'features' instead.

File: labgrid/environment.py [hunk 3]
```diff
@@ -43,6 +43,12 @@ def get_target(self, role: str = 'main') -> Target:
 
         return self.targets[role]
 
+    def get_feature_flags(self):
+        return self.config.get_flags()
+
+    def get_target_feature_flags(self, name):
+        return self.config.get_targets().get(name).get('feature flags')
```
Comment (jluebbe) [inline labgrid/environment.py]: No spaces here as well.

File: labgrid/environment.py [hunk 4]
```diff
@@ -43,6 +43,12 @@ def get_target(self, role: str = 'main') -> Target:
 
         return self.targets[role]
 
+    def get_feature_flags(self):
+        return self.config.get_flags()
```
Comment (jluebbe) [inline labgrid/environment.py]: This is broken at this point in the series. Also, 'environment: return flags for all targets' changes this code completely.

File: tests/test_fixtures.py [hunk 5]
```diff
@@ -43,12 +43,14 @@ def test_env_old_fixture(short_env, short_test):
         assert spawn.exitstatus == 0
 
 def test_env_env_fixture(short_env, short_test):
-    env=os.environ.copy()
-    env['LG_ENV'] = short_env
+    env=os.environ
+    env['LG_ENV'] = str(short_env)
+    print(env)
```
Comment (jluebbe) [inline tests/test_fixtures.py]: Leftover debug print? Please also describe what was broken and how this change fixes the problem.

File: labgrid/pytestplugin/reporter.py [hunk 6]
```diff
@@ -33,6 +35,26 @@ def pytest_configure(config):
         logging.getLogger().setLevel(logging.DEBUG)
     if lg_log:
         ConsoleLoggingReporter(lg_log)
+    env_config = config.option.env_config
```
Comment (jluebbe) [inline labgrid/pytestplugin/reporter.py]: Please describe why this change is needed.

File: labgrid/environment.py [hunk 7]
```diff
@@ -44,10 +44,15 @@ def get_target(self, role: str = 'main') -> Target:
...
-        return self.config.get_targets().get(name).get('feature flags')
+        return self.config.get_feature_flags()
+
+    def get_target_feature_flags(self):
+        flags = []
+        for target, value in self.config.get_targets().items():
+            flags += value.get('feature flags', [])
+            if flags:
+                flags = list(set(flags))
+        return flags
```
Comment (jluebbe) [inline labgrid/environment.py]: Why not keep the flags as a set from the beginning (and also return a set)?

File: doc/usage.rst [hunk 8]
```diff
@@ -296,6 +296,52 @@ For this example, you should get a report similar to this:
...
+Feature Flags
+~~~~~~~~~~~~~
+Labgrid includes support for feature flags on a global and target scope.
+They will be concatenated and compared to a pytest mark on the test to decide
+whether this combination does support the flags.::
+
+   import pytest
+
+   @pytest.mark.lg_feature("camera")
+   def test camera(target):
```
Comment (jluebbe) [inline doc/usage.rst]: Replace ' ' in the function name with '_'.

File: doc/usage.rst [hunk 9]
```diff
@@ -296,6 +296,52 @@ For this example, you should get a report similar to this:
 
   ========================== 3 passed in 29.77 seconds ===========================
 
+Feature Flags
+~~~~~~~~~~~~~
+Labgrid includes support for feature flags on a global and target scope.
+They will be concatenated and compared to a pytest mark on the test to decide
+whether this combination does support the flags.::
```
Comment (jluebbe) [inline doc/usage.rst]: "whether the test can run with the available features"?

**Comment by Emantor** [comment]
I pushed a new version which should address your review comments.

I merged several environments commits and dropped the fix for the environment fixture test.

File: labgrid/environment.py (hunk 7 above)
Comment (jluebbe) [reply labgrid/environment.py]: This is not yet addressed. Also, the commit message no longer matches the method names.

File: labgrid/config.py [hunk 10]
```diff
@@ -196,3 +196,6 @@ def get_images(self):
             images[name] = self.resolve_path(image)
 
         return images
+
+    def get_features(self):
```
Comment (jluebbe) [inline labgrid/config.py]: The commit message doesn't match the method name.

File: labgrid/pytestplugin/reporter.py [hunk 11]
```diff
@@ -21,6 +21,40 @@ def bold(text):
...
+
+    have_feature = set(env.get_features() + env.get_target_features())
+
+    for item in items:
+        want_feature = []
+        marker = item.get_marker("lg_feature")
+        if marker:
+            arg = marker.args[0]
+            if isinstance(arg, str):
+                want_feature.append(arg)
```
Comment (jluebbe) [inline labgrid/pytestplugin/reporter.py]: As you're not looping here, you can just assign want_feature directly. No need to create it as a empty list first.

File: labgrid/pytestplugin/reporter.py [hunk 12]
```diff
@@ -21,6 +21,40 @@ def bold(text):
...
+            elif isinstance(arg, list):
+                want_feature += arg
+            else:
+                raise Exception("Unsupported feature argument type")
+        missing_feature = []
+        for feature in want_feature:
+            if feature not in have_feature:
+                missing_feature.append(feature)
```
Comment (jluebbe) [inline labgrid/pytestplugin/reporter.py]: You can simplify this to set operations.
```
>>> {1,2,3} - {1,2}
{3}
```

File: labgrid/pytestplugin/reporter.py [hunk 13]
```diff
@@ -21,6 +21,40 @@ def bold(text):
...
+        if len(missing_feature) > 0:
```
Comment (jluebbe) [inline labgrid/pytestplugin/reporter.py]: Drop the len().

File: labgrid/pytestplugin/reporter.py [hunk 14]
```diff
@@ -21,6 +21,40 @@ def bold(text):
 def under(text):
     return "\033[4m{}\033[0m".format(text)
 
+@pytest.hookimpl()
+def pytest_collection_modifyitems(config, items):
```
Comment (jluebbe) [inline labgrid/pytestplugin/reporter.py]: As this has nothing to do with the reporter, this should go into a separate file. Maybe `collection.py`?

Comment (Emantor) [reply labgrid/pytestplugin/reporter.py]: Maybe its a good idea to move the `pytest_configure` function as well into the __init__.py?

File: labgrid/pytestplugin/reporter.py (hunk 11 above)
Comment (Emantor) [reply labgrid/pytestplugin/reporter.py]: I need to initialize the variable, otherwise I'll get an error while calculating the missing features.

File: labgrid/pytestplugin/reporter.py (hunk 12 above)
Comment (Emantor) [reply labgrid/pytestplugin/reporter.py]: Done.

File: labgrid/config.py (hunk 10 above)
Comment (Emantor) [reply labgrid/config.py]: Fixed.

File: labgrid/pytestplugin/reporter.py (hunk 13 above)
Comment (Emantor) [reply labgrid/pytestplugin/reporter.py]: Fixed.

**Comment by Emantor** [comment]
Another update:
- [x] fix usage rst examples
- [x] move to set operations for the feature calculation
- [x] operate on sets in the environment
Please review again.
